File: client-sdks/sockudo-js/test/forward-compat.test.ts
import { readFileSync } from "node:fs";
import { encode as encodeMsgpack } from "@msgpack/msgpack";
import { describe, expect, it, beforeAll, afterEach, vi } from "vitest";

import Channel from "../src/core/channels/channel";
import PresenceChannel from "../src/core/channels/presence_channel";
import Protocol from "../src/core/connection/protocol/protocol";
import { setProtocolVersion } from "../src/core/protocol_prefix";
import { setWireFormat } from "../src/core/wire_format";

const fixtureNames = [
  "future-v2-frame.json",
  "future-versioned-action.json",
  "future-webhook-events.json",
  "unknown-ai-extras.json",
] as const;

function fixtureRaw(name: (typeof fixtureNames)[number]): string {
  return readFileSync(
    new URL(`../../../tests/ai-conformance/fixtures/forward-compat/${name}`, import.meta.url),
    "utf8",
  );
}

function decodeRaw(raw: string) {
  return Protocol.decodeMessage({ data: raw } as MessageEvent);
}

function replayRaw(channel: Channel, raw: string) {
  const event = decodeRaw(raw);
  if (typeof event.event === "string" && event.event.length > 0 && event.channel === channel.name) {
    channel.handleEvent(event);
  }
  return event;
}

beforeAll(() => {
  Object.assign(globalThis, {
    VERSION: "test-version",
    CDN_HTTP: "",
    CDN_HTTPS: "",
    DEPENDENCY_SUFFIX: "",
  });
});

afterEach(() => {
  vi.restoreAllMocks();
  setProtocolVersion(7);
  setWireFormat("json");
});

describe("forward-compatible realtime fixture replay", () => {
  it("decodes and replays forward-compat frames without corrupting channel state", () => {
    setProtocolVersion(2);
    setWireFormat("json");
    const channel = new Channel("private-ai-forward", {
      connection: { socket_id: "1.1", state: "connected" },
      send_event: vi.fn(),
      unsubscribe: vi.fn(),
    } as any);
    const known = vi.fn();
    const global = vi.fn();
    channel.bind("app-known", known);
    channel.bind_global(global);

    channel.handleEvent({
      event: "app-known",
      channel: "private-ai-forward",
      data: { before: true },
    } as any);
    const beforeState = {
      subscribed: channel.subscribed,
      subscriptionPending: channel.subscriptionPending,
      subscriptionCancelled: channel.subscriptionCancelled,
    };

    const decoded = fixtureNames.map((name) => replayRaw(channel, fixtureRaw(name)));

    channel.handleEvent({
      event: "app-known",
      channel: "private-ai-forward",
      data: { after: true },
    } as any);

    expect(known).toHaveBeenCalledTimes(2);
    expect({
      subscribed: channel.subscribed,
      subscriptionPending: channel.subscriptionPending,
      subscriptionCancelled: channel.subscriptionCancelled,
    }).toEqual(beforeState);
    expect(decoded.map((event) => event.event)).toEqual([
      "sockudo:future_event",
      "sockudo:message.future",
      undefined,
      "ai-output",
    ]);
    expect(global).toHaveBeenCalledWith("sockudo:future_event", {
      known: true,
      futureObject: { nested: "ignored" },
    });
  });

  it("preserves extras.ai and safe serials across decode boundaries", () => {
    setProtocolVersion(2);
    setWireFormat("json");

    const futureFrame = decodeRaw(fixtureRaw("future-v2-frame.json"));
    const extrasFrame = decodeRaw(fixtureRaw("unknown-ai-extras.json"));
    const safeFrame = decodeRaw(
      JSON.stringify({
        event: "sockudo:test",
        channel: "private-ai-forward",
        serial: 2147483648,
      }),
    );

    expect(futureFrame.serial).toBe("9007199254740993");
    expect(safeFrame.serial).toBe(2147483648);
    expect(extrasFrame.extras?.ai).toEqual({
      transport: {
        "turn-id": "turn-1",
        status: "streaming",
      },
      codec: {
        "provider-future-key": "opaque",
        "x-custom": "opaque",
      },
    });
    expect(extrasFrame.extras?.futureExtrasField).toBe(true);
  });

  it("normalizes binary uint64 serials without BigInt leakage", () => {
    setProtocolVersion(2);
    setWireFormat("messagepack");
    const msgpackPayload = encodeMsgpack(
      [
        "sockudo:test",
        "private-ai-forward",
        ["string", "content"],
        null,
        null,
        null,
        null,
        null,
        null,
        "stream-1",
        BigInt("9007199254740993"),
        null,
        {
          headers: {
            sockudo_history_serial: ["number", BigInt("9007199254740994")],
          },
          ai: {
            transport: { "turn-id": "turn-1" },
          },
        },
      ],
      { useBigInt64: true },
    );

    const decoded = Protocol.decodeMessage({ data: msgpackPayload } as MessageEvent);

    expect(decoded.serial).toBe("9007199254740993");
    expect(decoded.extras?.headers?.sockudo_history_serial).toBe("9007199254740994");
    expect(decoded.extras?.ai).toEqual({
      transport: { "turn-id": "turn-1" },
    });
  });

  it("routes unknown internal presence events without changing the member map", () => {
    setProtocolVersion(2);
    const channel = new PresenceChannel("presence-ai-forward", {
      connection: { socket_id: "1.1", state: "connected" },
      user: { signinDonePromise: Promise.resolve(), user_data: null },
      config: {},
      send_event: vi.fn(),
      unsubscribe: vi.fn(),
    } as any);
    const internal = vi.fn();
    channel.bind_global(internal);
    channel.members.setMyID("user-1");
    channel.members.onSubscription({
      presence: {
        hash: { "user-1": { role: "reader" } },
        count: 1,
      },
    });

    channel.handleEvent({
      event: "sockudo_internal:presence_update",
      channel: "presence-ai-forward",
      data: { user_id: "user-1", user_info: { role: "writer" } },
    } as any);
    channel.handleEvent({
      event: "sockudo_internal:member_added",
      channel: "presence-ai-forward",
      data: { user_info: { malformed: true } },
    } as any);

    expect(internal).toHaveBeenCalledWith("sockudo_internal:presence_update", {
      user_id: "user-1",
      user_info: { role: "writer" },
    });
    expect(channel.members.count).toBe(1);
    expect(channel.members.get("user-1")).toEqual({
      id: "user-1",
      info: { role: "reader" },
    });
    expect(channel.members.get("undefined")).toBeNull();
  });
});

File: client-sdks/sockudo-kotlin/lib/src/test/kotlin/io/sockudo/client/ForwardCompatTest.kt
package io.sockudo.client

import java.lang.reflect.InvocationTargetException
import java.nio.file.Files
import java.nio.file.Path
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ForwardCompatTest {
    @Test
    fun replaysForwardCompatFixturesThroughClientDispatchWithoutCorruptingChannelState() {
        val client = testClient(connectionRecovery = true)
        val channel = client.subscribe("private-ai-forward")
        val knownEvents = mutableListOf<Any?>()
        val globalEvents = mutableListOf<Pair<String, Any?>>()
        val errors = mutableListOf<Any?>()

        channel.bind("app-known") { data, _ -> knownEvents += data }
        channel.bindGlobal { eventName, data -> globalEvents += eventName to data }
        client.bind("error") { data, _ -> errors += data }

        replayRaw(client, """{"event":"app-known","channel":"private-ai-forward","data":{"before":true}}""")
        val beforeState =
            Triple(
                channel.isSubscribed,
                channel.subscriptionPending,
                channel.subscriptionCancelled,
            )

        fixtureNames.forEach { name -> replayRaw(client, fixtureRaw(name)) }

        replayRaw(client, """{"event":"app-known","channel":"private-ai-forward","data":{"after":true}}""")

        assertEquals(2, knownEvents.size)
        assertEquals(
            beforeState,
            Triple(
                channel.isSubscribed,
                channel.subscriptionPending,
                channel.subscriptionCancelled,
            ),
        )
        assertEquals(9_007_199_254_740_993L, client.getRecoveryPosition("private-ai-forward")?.serial)
        assertTrue(globalEvents.any { it.first == "sockudo:future_event" })
        assertTrue(globalEvents.any { it.first == "sockudo:message.future" })
        assertTrue(globalEvents.any { it.first == "ai-output" })
        assertEquals(1, errors.count { it is SockudoException.MessageParseError })
    }

    @Test
    fun preservesAiAndUnknownExtrasFromForwardCompatFixture() {
        val decoded = ProtocolCodec.decodeEvent(fixtureRaw("unknown-ai-extras.json"), SockudoWireFormat.json)
        val transport = decoded.extras?.ai?.get("transport") as? Map<*, *>
        val codec = decoded.extras?.ai?.get("codec") as? Map<*, *>

        assertEquals("turn-1", transport?.get("turn-id"))
        assertEquals("streaming", transport?.get("status"))
        assertEquals("opaque", codec?.get("provider-future-key"))
        assertEquals(true, decoded.extras?.raw?.get("futureExtrasField"))
        assertEquals(true, decoded.extras?.get("futureExtrasField"))
    }

    @Test
    fun decodesSerialBoundariesWithoutIntOrDoubleTruncation() {
        val overInt = 2_147_483_648L
        val overSafe = 9_007_199_254_740_993L

        val jsonOverInt =
            ProtocolCodec.decodeEvent(
                """{"event":"sockudo:test","channel":"private-ai-forward","serial":$overInt}""",
                SockudoWireFormat.json,
            )
        val jsonOverSafe =
            ProtocolCodec.decodeEvent(
                """{"event":"sockudo:test","channel":"private-ai-forward","serial":"$overSafe"}""",
                SockudoWireFormat.json,
            )
        val msgpackOverSafe =
            ProtocolCodec.decodeEvent(
                ProtocolCodec.encodeEnvelope(
                    linkedMapOf(
                        "event" to "sockudo:test",
                        "channel" to "private-ai-forward",
                        "serial" to overSafe.toString(),
                    ),
                    SockudoWireFormat.messagepack,
                ),
                SockudoWireFormat.messagepack,
            )
        val protobufOverSafe =
            ProtocolCodec.decodeEvent(
                ProtocolCodec.encodeEnvelope(
                    linkedMapOf(
                        "event" to "sockudo:test",
                        "channel" to "private-ai-forward",
                        "stream_id" to "stream-1",
                        "serial" to overSafe,
                    ),
                    SockudoWireFormat.protobuf,
                ),
                SockudoWireFormat.protobuf,
            )

        assertEquals(overInt, jsonOverInt.serial)
        assertEquals(overSafe, jsonOverSafe.serial)
        assertEquals(overSafe, msgpackOverSafe.serial)
        assertEquals(overSafe, protobufOverSafe.serial)
        assertEquals("stream-1", protobufOverSafe.streamId)
    }

    @Test
    fun parsesVersionedIntegerStringHeadersExactlyAndIgnoresUnknownActions() {
        val event =
            SockudoEvent(
                event = "sockudo:message.update",
                rawMessage = "{}",
                extras =
                    MessageExtras(
                        headers =
                            mapOf(
                                "sockudo_action" to "message.update",
                                "sockudo_message_serial" to "msg-1",
                                "sockudo_history_serial" to "9007199254740993",
                                "sockudo_version_timestamp_ms" to "1710000000000",
                            ),
                    ),
            )
        val futureAction =
            event.copy(
                event = "sockudo:message.future",
                extras =
                    MessageExtras(
                        headers =
                            mapOf(
                                "sockudo_action" to "message.future",
                                "sockudo_message_serial" to "msg-1",
                                "sockudo_history_serial" to "9007199254740993",
                            ),
                    ),
            )

        val info = getMutableMessageInfo(event)

        assertEquals(MutableMessageAction.UPDATE, info?.action)
        assertEquals(9_007_199_254_740_993L, info?.historySerial)
        assertEquals(1_710_000_000_000L, info?.versionTimestampMs)
        assertEquals(null, getMutableMessageInfo(futureAction))
    }

    @Test
    fun presenceUpdatesApplyWithoutLettingMalformedMemberPayloadsMutateMembers() {
        val client = testClient()
        val channel = client.subscribe("presence-ai-forward") as PresenceChannel
        val emittedEvents = mutableListOf<Pair<String, Any?>>()
        channel.bindGlobal { eventName, data -> emittedEvents += eventName to data }
        channel.members.rememberMyId("user-1")
        channel.handle(
            SockudoEvent(
                event = client.p.internal_("subscription_succeeded"),
                channel = "presence-ai-forward",
                data =
                    mapOf(
                        "presence" to
                            mapOf(
                                "hash" to mapOf("user-1" to mapOf("role" to "reader")),
                                "count" to 1,
                            ),
                    ),
                rawMessage = "{}",
            ),
        )

        channel.handle(
            SockudoEvent(
                event = client.p.internal_("presence_update"),
                channel = "presence-ai-forward",
                data = mapOf("user_id" to "user-1", "user_info" to mapOf("role" to "writer")),
                rawMessage = "{}",
            ),
        )
        channel.handle(
            SockudoEvent(
                event = client.p.internal_("member_added"),
                channel = "presence-ai-forward",
                data = mapOf("user_info" to mapOf("malformed" to true)),
                rawMessage = "{}",
            ),
        )
        channel.handle(
            SockudoEvent(
                event = client.p.internal_("member_removed"),
                channel = "presence-ai-forward",
                data = listOf("malformed"),
                rawMessage = "{}",
            ),
        )

        assertEquals(1, channel.members.count)
        assertEquals(mapOf("role" to "writer"), channel.members.member("user-1")?.info)
        assertEquals(null, channel.members.member("undefined"))
        assertTrue(emittedEvents.any { it.first == client.p.event("presence_update") })
    }

    private fun testClient(connectionRecovery: Boolean = false): SockudoClient =
        SockudoClient(
            "app-key",
            SockudoOptions(
                cluster = "local",
                forceTls = false,
                enabledTransports = listOf(SockudoTransport.ws),
                wsHost = "127.0.0.1",
                wsPort = 6001,
                connectionRecovery = connectionRecovery,
            ),
        )

    private fun replayRaw(
        client: SockudoClient,
        raw: String,
    ) {
        val method = SockudoClient::class.java.getDeclaredMethod("handleRawMessage", Any::class.java)
        method.isAccessible = true
        try {
            method.invoke(client, raw)
        } catch (error: InvocationTargetException) {
            throw error.targetException
        }
    }

    private fun fixtureRaw(name: String): String {
        val candidates =
            listOf(
                Path.of("..", "..", "tests", "ai-conformance", "fixtures", "forward-compat", name),
                Path.of("..", "..", "..", "tests", "ai-conformance", "fixtures", "forward-compat", name),
                Path.of("tests", "ai-conformance", "fixtures", "forward-compat", name),
            )
        val path = candidates.firstOrNull { Files.exists(it) }
            ?: error("Unable to locate forward-compat fixture '$name'")
        return Files.readString(path)
    }

    private companion object {
        val fixtureNames =
            listOf(
                "future-v2-frame.json",
                "future-versioned-action.json",
                "future-webhook-events.json",
                "unknown-ai-extras.json",
            )
    }
}
